soc: st: stm32: common: fix wakeup off-by-one

rbrscoleman opened the pull request:
Wakeup pin indices are 1-based, but `LISTIFY` is 0-based. Consequently, the last element of the pin-to-register-bit lookup table was never populated.

For example, STM32L4xx MCUs, like the STM32L4R5ZI on the NUCLEO-L4R5ZI board supported by the `samples/boards/st/power_mgmt/wkup_pins` sample, [have five WKUP pins](https://github.com/zephyrproject-rtos/zephyr/blob/v4.0.0/dts/arm/st/l4/stm32l4.dtsi#L486-L511), numbered 1 through 5. However, the `LISTIFY` invocation populating `table_wakeup_pins` ranged over [0, 4], and so the table was actually defined as:

```c
static const uint32_t table_wakeup_pins[6] = {
	LL_PWR_WAKEUP_PIN0, LL_PWR_WAKEUP_PIN1, LL_PWR_WAKEUP_PIN2, LL_PWR_WAKEUP_PIN3, LL_PWR_WAKEUP_PIN4,};
```

I.e., the last member was 0-filled (and the first member – `LL_PWR_WAKEUP_PIN0` – is synthetic, and is defined in the adjacent header only to suppress an error generating this array). Attempts to actually configure WKUP5 would fail silently.

Happy to hear feedback on other ways to accomplish this fix – it is quite verbose.

mathieuchopstm commented on soc/st/stm32/common/stm32_wkup_pins.c:
```diff
@@ -133,7 +131,31 @@ struct wkup_pin_cfg_t {
  */
 static const uint32_t table_wakeup_pins[PWR_STM32_MAX_NB_WKUP_PINS + 1] = {
-	LISTIFY(PWR_STM32_MAX_NB_WKUP_PINS, PWR_STM32_WKUP_PIN_DEF, (,))};
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 1 && defined(LL_PWR_WAKEUP_PIN1)
+	[1] = LL_PWR_WAKEUP_PIN1,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 2 && defined(LL_PWR_WAKEUP_PIN2)
+	[2] = LL_PWR_WAKEUP_PIN2,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 3 && defined(LL_PWR_WAKEUP_PIN3)
+	[3] = LL_PWR_WAKEUP_PIN3,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 4 && defined(LL_PWR_WAKEUP_PIN4)
+	[4] = LL_PWR_WAKEUP_PIN4,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 5 && defined(LL_PWR_WAKEUP_PIN5)
+	[5] = LL_PWR_WAKEUP_PIN5,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 6 && defined(LL_PWR_WAKEUP_PIN6)
+	[6] = LL_PWR_WAKEUP_PIN6,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 7 && defined(LL_PWR_WAKEUP_PIN7)
+	[7] = LL_PWR_WAKEUP_PIN7,
+#endif
+#if PWR_STM32_MAX_NB_WKUP_PINS >= 8 && defined(LL_PWR_WAKEUP_PIN8)
+	[8] = LL_PWR_WAKEUP_PIN8,
+#endif
+};
```
`IF_ENABLED` can be used inside macros as replacement for `#if defined()`, which allows us to put the check in a macro that can be `LISTIFY`'ed.

For the off-by-one, I think it's reasonable to have a dummy value at index 0, before the `LISTIFY`.

Rough idea of what this could look like:
```c
/* Keep PWR_STM32_WKUP_PIN_DEF defined (although it could use _CONCAT instead) */

/* To be discussed: are discontinuous ranges a potential issue? (e.g., PIN3 exists but not PIN2)
 * If yes, COND_CODE_1 should be used instead.
 */
#define PIN_IF_EXISTS(x) \
    IF_ENABLED(PWR_STM32_WKUP_PIN_DEF(x), (PWR_STM32_WKUP_PIN_DEF(x)))

/* LISTIFY is [0..n-1] but we care about [1..n]
#define LISTIFY_WKUP_PIN_HELPER(x) \
        PIN_IF_EXISTS(UTIL_INC(x))
```

then the whole array can be replaced with a simple:
```suggestion
  0, /* Dummy entry - PIN0 does not exist */
  LISTIFY(PWR_STM32_MAX_NB_WKUP_PINS, LISTIFY_WKUP_PIN_HELPER, (,))
};
```

rbrscoleman replied:
Thank you for looking at this!

Unless I'm missing something, that still doesn't populate a valid array member at `table_wakeup_pins[PWR_STM32_MAX_NB_WKUP_PINS]`, because `LISTIFY` ranges over [0, limit). E.g., with `PWR_STM32_MAX_NB_WKUP_PINS == 5`, you'll get an array with `{0, LL_PWR_WAKEUP_PIN1, LL_PWR_WAKEUP_PIN2, LL_PWR_WAKEUP_PIN3, LL_PWR_WAKEUP_PIN4}`.

> are discontinuous ranges a potential issue? (e.g., PIN3 exists but not PIN2)

Just based on one of the STM32 LL headers (`stm32g0xx_ll_pwr.h`), it does appear to be possible to have a STM32G0 with WKUP4 but not WKUP3.

rbrscoleman replied:
Ah, sorry, I missed `UTIL_INC`! I'll try this.

mathieuchopstm replied:
> > are discontinuous ranges a potential issue? (e.g., PIN3 exists but not PIN2)
> 
> Just based on one of the STM32 LL headers (`stm32g0xx_ll_pwr.h`), it does appear to be possible to have a STM32G0 with WKUP4 but not WKUP3.

Indeed. This also appears to be the case for STM32WB series.

Since the corresponding `wkup-pin` DTS nodes should not be active on these series, it sounds safe to have the corresponding indexes in the array be zero: in this case, the following using `COND_CODE_1` (as I suggested) should work:
```c
#define PIN_IF_EXISTS(x) \
    COND_CODE_1(PWR_STM32_WKUP_PIN_DEF(x),
         (PWR_STM32_WKUP_PIN_DEF(x)),
         (0)
    )
```

gautierg-st commented on soc/st/stm32/common/stm32_wkup_pins.c:
```diff
@@ -133,7 +131,31 @@ struct wkup_pin_cfg_t {
  * @brief LookUp Table to store LL_PWR_WAKEUP_PINx for each wake-up pin.
  */
 static const uint32_t table_wakeup_pins[PWR_STM32_MAX_NB_WKUP_PINS + 1] = {
-	LISTIFY(PWR_STM32_MAX_NB_WKUP_PINS, PWR_STM32_WKUP_PIN_DEF, (,))};
```
I think LISTIFY could still be used by fixing it like this:
```
#define PWR_STM32_WKUP_PIN_DEF(i, _) _CONCAT1(LL_PWR_WAKEUP_PIN, UTIL_INC(i))
static const uint32_t table_wakeup_pins[PWR_STM32_MAX_NB_WKUP_PINS + 1] = {
	0,
	LISTIFY(PWR_STM32_MAX_NB_WKUP_PINS, PWR_STM32_WKUP_PIN_DEF, (,))};
```
I've done something similar here https://github.com/zephyrproject-rtos/zephyr/pull/82955/commits/c29a8b71b37e1db6d605b546cd3303703d2d5b9d

rbrscoleman committed:
soc: st: stm32: common: fix wakeup off-by-one.

Wakeup pin indices are 1-based, but `LISTIFY` is 0-based. Consequently, the
last element of the pin-to-register-bit lookup table was never populated.

Signed-off-by: Samuel Coleman <samuel.coleman@rbr-global.com>

rbrscoleman commented:
Thank you both for your feedback. I've revised to basically what Guillaume suggested.

Unfortunately, I was not able to get a `COND_CODE_1`/`IS_ENABLED`/`IF_ENABLED`-based existence check for pin definitions to work, because all of those depend on the macro in question being defined as “1”. Because the `LL_PWR_WAKEUP_PIN1x` macros define bit masks, that condition only holds for the first one. Happy to take another stab at this if I've missed something else!

gautierg-st commented:
Maybe you can use the macro IS_EMPTY. It checks if the macro has a value (any value, not just 1).
But if you're fed up with Zephyr macros, that's perfectly ok, we can merge the current fix ;)

rbrscoleman commented:
Unfortunately, I can't get that to work, either, as it requires the macro under test to be defined – `IS_EMPTY()` returns false for undefined macros, which puts me right back where I started. I guess I could change `stm32_wkup_pins.h` to define missing pins as empty rather than 0, but I don't really believe that would be a meaningful improvement.

I think I'm personally happy with this as it stands – tremendously less verbose than it was, thank you – and it does meet our internal needs.

gautierg-st reviewed (approved):
Well if it doesn't work without defining the missing PIN, I agree to stop here. Current PR is already a nice improvement.
Thanks!

JarmouniA reviewed (approved):
Thanks for the fix :)
